### PR TITLE
Add Sync Subnets Subscribers

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Altair Work

**What does this PR do? Why is it needed?**

- [x] Builds on from #9538 
- [x] Adds in the sync subnet subscriber methods. 
- [x] Adds in routines to dynamically/statically select for sync subnets.

**Which issues(s) does this PR fix?**

Part of #8638 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/sync/metrics.go [hunk 1]
```diff
@@ -20,6 +22,12 @@ var (
 			Help: "The number of peers subscribed to a given topic.",
 		}, []string{"topic"},
 	)
+	subscribedTopicPeerCount = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "p2p_subscribed_topic_peer_count",
```
Comment (terencechain) [inline beacon-chain/sync/metrics.go:27]:
```suggestion
			Name: "p2p_subscribed_topic_peer_total",
```

File: beacon-chain/sync/service.go [hunk 2]
```diff
@@ -252,7 +252,12 @@ func (s *Service) registerHandlers() {
 					return
 				}
 				// Register respective pubsub handlers at state synced event.
-				s.registerSubscribers()
+				digest, err := s.currentForkDigest()
+				if err != nil {
+					log.WithError(err).Error("Could not retrieve current fork digest")
```
Comment (terencechain) [inline beacon-chain/sync/service.go:257]: don't you want to return if we could not retrieve fork digest here?

**Review by terencechain — CHANGES_REQUESTED** [review]

File: beacon-chain/sync/subscriber_sync_contribution_proof.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"google.golang.org/protobuf/proto"
+)
+
+// syncContributionAndProofSubscriber forwards the incoming validated sync contributions and proof to the
+// contribution pool for processing.
+// skipcq: SCC-U1000
+func (s *Service) syncContributionAndProofSubscriber(_ context.Context, msg proto.Message) error {
+	a, ok := msg.(*ethpb.SignedContributionAndProof)
```
Comment (terencechain) [inline beacon-chain/sync/subscriber_sync_contribution_proof.go:16]: `a` ? I think this was copied from attestation

File: beacon-chain/sync/subscriber.go [hunk 4]
```diff
@@ -412,6 +438,155 @@ func (s *Service) subscribeAggregatorSubnet(
...
+					log.Error(err)
+					continue
+				}
+				if !valid {
+					log.Warnf("Sync subnets with digest %#x are no longer valid, unsubscribing from all of them.", digest)
+					// Unsubscribes from all our current subnets.
+					s.reValidateSubscriptions(subscriptions, []uint64{}, topicFormat, digest)
+					ticker.Done()
+					return
+				}
```
Comment (terencechain) [inline beacon-chain/sync/subscriber.go:575]: Im don't see why we need to put this in the for select loop. `digest, genesis, genRoot` doesn't change
Wouldn't it make more sense to put this validation outside of go routine and short circuit the exit early

File: beacon-chain/sync/subscriber.go [hunk 5]
```diff
@@ -412,6 +438,155 @@ func (s *Service) subscribeAggregatorSubnet(
...
+		log.Debugf("No peers found subscribed to sync gossip subnet with "+
+			"committee index %d. Searching network for peers subscribed to the subnet.", idx)
+		_, err := s.cfg.P2P.FindPeersWithSubnet(s.ctx, subnetTopic, idx, params.BeaconNetworkConfig().MinimumPeersInSubnet)
+		if err != nil {
+			log.WithError(err).Debug("Could not search for peers")
+		}
+	}
+}
+
+// subscribe to a static subnet  with the given topic and index.A given validator and subscription handler is
```
Comment (terencechain) [inline beacon-chain/sync/subscriber.go:467]:
```suggestion
// subscribe to a static subnet with the given topic and index. A given validator and subscription handler is
```

File: beacon-chain/sync/subscriber.go (hunk 4 above)
Comment (nisdas) [reply beacon-chain/sync/subscriber.go:575]: so `isDigestValid` checks if the provided digest is our current digest, if it is not we would exit this method. This is to handle unsubscribing at fork boundaries.

**Review by terencechain — APPROVED** [review]

**Pull request merged**